### PR TITLE
G-Loader

### PR DESCRIPTION
**Pull request by antoinechauvn**
Added G-Loader extension

### CODE REVIEW

**Comment by WiredSpast** [comment]
The extension.json file has to be a file in this repo, not a file in the zip

**Comment by WiredSpast** [comment]
The icon file is also to big, the size limit is 80x80px (it will be resized to 40px in the store, so best to use a 40x40px image)

**Comment by WiredSpast** [comment]
image
The files should also be directly in the root folder of the zip not in a subfolder (unless you edit the command to say `G-Loader/__main__.py` instead of `__main__.py`)

**Comment by WiredSpast** [comment]
image
The command has to include the cookie and file as well, so it should be
```json
"commands": {
  "default": ["python", "__main__.py", "-c", "{cookie}", "-p", "{port}", "-f", "{filename}"]
},
```

**Comment by antoinechauvn** [comment]
Updated

**Comment by WiredSpast** [comment]
You still forgot to add the `extension.json` file into this repo

**Comment by antoinechauvn** [comment]
Done

File: store/extensions/G-Loader/extension.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "categories": ["Usability"],
+
+    "source": "https://github.com/antoinechauvn/G-Loader",
+	"readme": "https://github.com/antoinechauvn/G-Loader/README.md",
+
+    "stable": true,
+
+    "framework": {
+      "name": "G-Python",
+      "version": "0.1.5"
```
Comment (WiredSpast) [inline store/extensions/G-Loader/extension.json]: You appear to have used 0.1.6 not 0.1.5
```suggestion
      "version": "0.1.6"

```

**Comment by WiredSpast** [comment]
The extension doesn't terminate when G-Earth does, it should
(This happens because the UI thread keeps running while the extension thread dies)

**Comment by WiredSpast** [comment]
You're extension seems to have an extra dependency next to G-Python, I recommend using PyExtensionSetup and a `requirements.txt` file, the extension setup will make sure the user has python and the necesarry requirements installed.

The launch command would look like this:
```json
"default": ["java", "-jar", "PyExtensionSetup.jar", "-e", "__main__.py", "-v", "3.2.0", "-c", "{cookie}", "-p", "{port}", "-f", "{filename}"]
```
and the requirements.txt would have this in it:
```
g-python==0.1.6
PySide6==6.4.2
```
and then include both the requirements.txt file and PyExtensionSetup.jar in your extension.zip

PyExtensionSetup release: https://github.com/WiredSpast/PyExtensionSetup/releases/latest

**Comment by antoinechauvn** [comment]
Done

**Comment by WiredSpast** [comment]
The extension still remains running after G-Earth closes, you can fix this by using a timer from the threading package which checks whether the extension's connection has been closed or not every second:

```py
# Check if extension is still connected to G-Earth every second
def checkExtensionActive():
  if ext.is_closed():
    # Close window and stop if extension isn't connected
    window.close(); # This is pseudo code I haven't looked at how to close a window in the framework you're using
        
    # Repeat every second
    Timer(1.0, checkExtensionActive).start()
    
checkExtensionActive()
```

I've also discussed the permanently open UI with sirjonas and we've come to an agreement that extensions in the store shouldn't have a permanently open UI. It should start hidden and be opened by the green playbutton, and the window should then be able to be closed again (either by use of the same playbutton or by an X button):

```py
# window.toggle is again pseudocode
ext.on_event('double_click', window.show)
```

The extension is also currently always on top, it shouldn't unless you've allowed a user to say that they want it to remain on top (using for a checkbox like G-Earth has in it's extra tab).

**Comment by antoinechauvn** [comment]
Updated

**Comment by antoinechauvn** [comment]
- Fixed Qt autostart due to G-Earth running all extensions files.
```python
def display():
    try:
        window = engine.rootObjects()[0]
    except:
        engine.load(qml_file)
    else:
        window.setVisible(not window.isVisible())
```

- Fixed Window size and close button
- Fixed Habbo.com Api maintenance

File: store/extensions/G-Loader/extension.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      "default": ["java", "-jar", "PyExtensionSetup.jar", "-e", "__main__.py", "-v", "3.2.0", "-c", "{cookie}", "-p", "{port}", "-f", "{filename}"]
+    },
+
+    "compatibility": {
+        "systems": ["Linux", "Windows", "Mac"],
+        "clients": ["Nitro"]
+    },
+
+    "submissionDate": "21-01-2023 21:05:00",
+    "updateDate": "21-01-2023 21:05:00",
```
Comment (WiredSpast) [inline store/extensions/G-Loader/extension.json]: Might be better to set these to today, otherwise your extension is gonna be low down in the store

**Comment by WiredSpast** [comment]
The code in your zip file currently doesn't match to the code in your repo, make sure these match

**Pull request merged**